src/include/frag.h

```diff
@@ -567,29 +567,27 @@ class fragset_t {
       f = f.parent();
     }
   }
-  
+
+  void insert_raw(frag_t f){
+    _set.insert(f);
+  }
   void insert(frag_t f) {
     _set.insert(f);
     simplify();
   }
 
   void simplify() {
-    while (1) {
-      bool clean = true;
-      std::set<frag_t>::iterator p = _set.begin();
-      while (p != _set.end()) {
-	if (!p->is_root() &&
-	    _set.count(p->get_sibling())) {
-	  _set.erase(p->get_sibling());
-	  _set.insert(p->parent());
-	  _set.erase(p++);
-	  clean = false;
-	} else {
-	  p++;
-	}
+    std::set<frag_t>::iterator p = _set.begin();
+    while (p != _set.end()) {
+      if (!p->is_root() &&
+	  _set.count(p->get_sibling())) {
+	_set.erase(p->get_sibling());
+	_set.insert(p->parent());
```

Comment:
Because these may rebalance the rbtree, you cannot necessarily do this in a single pass. You need to keep the `clean` variable and outer loop.

Reply:
updated